jobs,*: read job payload and progress from the job_info table

adityamaru opened the pull request:
This change touches all the parts of the code that were
relying on the `system.jobs` table to read the payload and/or
progress corresponding to a job. If the cluster version has advanced
past V23_1JobInfoTableIsBackfilled, every job in the system.jobs table
has a payload and progress entry written to the `job_info` table.
This change migrates callers to use this new table when reading the
payload and progress of a job.

The most important changes are in the logic that drives `crdb_internal.system_jobs`,
which in turn drives `crdb_internal.jobs` and `SHOW JOBS`. Additionally,
there are changes in how the registry resolves the progress and payload
when loading or resuming a job.

Several tests that read the payload and progress from `system.jobs` now
rely on `crdb_internal.system_jobs` to fetch this information from either
the jobs table or job_info table depending on the cluster version.

Fixes: https://github.com/cockroachdb/cockroach/issues/97762

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98608)
<!-- Reviewable:end -->

adityamaru committed:
jobs,*: read job payload and progress from the job_info table

This change touches all the parts of the code that were
relying on the `system.jobs` table to read the payload and/or
progress corresponding to a job. If the cluster version has advanced
past V23_1JobInfoTableIsBackfilled, every job in the system.jobs table
has a payload and progress entry written to the `job_info` table.
This change migrates callers to use this new table when reading the
payload and progress of a job.

The most important changes are in the logic that drives `crdb_internal.system_jobs`,
which in turn drives `crdb_internal.jobs` and `SHOW JOBS`. Additionally,
there are changes in how the registry resolves the progress and payload
when loading or resuming a job.

Several tests that read the payload and progress from `system.jobs` now
rely on `crdb_internal.system_jobs` to fetch this information from either
the jobs table or job_info table depending on the cluster version.

Fixes: #97762

Release note: None

miretskiy commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -733,7 +733,8 @@ func TestChangefeedCursor(t *testing.T) {
 		if e, ok := fooLogical.(cdctest.EnterpriseTestFeed); ok {
 			var bytes []byte
-			sqlDB.QueryRow(t, `SELECT payload FROM system.jobs WHERE id=$1`, e.JobID()).Scan(&bytes)
+			sqlDB.QueryRow(t, `SELECT payload FROM "".crdb_internal.system_jobs WHERE id = $1`,
```
curious: why do we need "" ?

adityamaru replied:
Added a comment to `jobs_verification.go` and changed most test places to use that query instead.

miretskiy commented on pkg/jobs/jobs.go:
```diff
@@ -750,6 +752,78 @@ func HasJobNotFoundError(err error) bool {
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "failed to get payload for job %d", j.ID())
+		}
+		if !exists {
```
do we intend on making progress optional at some point?

adityamaru replied:
_maybe_. On job creation we expect the job record to have a valid, non-null progress and payload so unless the user manually deletes the progress for a job from `system.job_info` we will always have an entry. That said `SHOW JOBS` goes out of its way to handle the absence of a progress row for a job. So I dunno we should decide if having progress is mandatory or not and then apply that through the system.

miretskiy commented on pkg/jobs/adopt.go:
```diff
@@ -144,10 +146,19 @@ COALESCE(last_run, created) + least(
 		` created_by_type, created_by_id  FROM system.jobs, ` + canRunArgs + " WHERE " + resumeQueryWhereBase
+
+	// deprecatedResumeQueryBaseCols loads the payload and progress from
+	// system.jobs instead of the system.job_info table.
```
can you add a todo or some such to deprecated this after release XXX?

miretskiy reviewed (approved):
I really only have nits.

dt reviewed (approved):
Nice work!

adityamaru commented:
`TestDecommission` is a flake - https://github.com/cockroachdb/cockroach/issues/96630
The other stress failure is during test server startup: `optInToDiagnosticsStatReporting: setting updated but timed out waiting to read new value` so nothing seems wrong with the test per-se. Will investigate separately.

adityamaru commented:
bors retry
